runtime/lua/vim/treesitter/dev.lua

```diff
@@ -595,7 +595,7 @@ function M.edit_query(lang)
   end
   vim.cmd(cmd)
 
-  local parser = vim.treesitter._get_parser(buf, lang)
+  local parser = vim.treesitter.get_parser(buf, lang, { error = false })
   if not parser then
     return nil, 'No parser available for the given buffer'
   end
```

Comment:
same?

Reply:
(sorry for the iterations; I want to get this right, since it'll establish a pattern for use elsewhere, e.g., my `add()` PR.)

Reply:
Concretely, my proposal is 'do not return an error message here, `nil` already means "No parser available", just return'.

Reply:
How else will the caller know why the function failed? Maybe there is only a single error path now, but there could be more in the future, and in that case we would want to explain what the error was. 

Returning an error message also lets callers simply use `local blah = assert(vim.treesitter.dev.show_tree())` which will conveniently show the error message “for free” (this is the same pattern Lua itself uses in e.g. the `io` library. I don’t see any reason to deviate here).

Reply:
Then that error message should provide more value, if it's the (only) one that is surfaced to the user. (At the very least, _which_ parser? _which_ buffer?)

Reply:
Also, I'm a bit confused, since we're commenting here on `:EditQuery`, which does not have a "caller" (in the sense of a calling function -- unless we are talking about the `nvim_create_usercommand`, which then needs to `assert` the call).

Sorry for being so ornery, but I'm genuinely interested in having a discussion about this so I can understand the _full_ pattern across our API (at least treesitter, possibly beyond). Ideally, we'll get something we can actually write in `develop.txt` and I can follow in future PRs instead of winging it based on contributor style as we (certainly I) have so far.

(Part of the concern is that adding the error message to the return value locks the return part of the API. If we want to promote this pattern, we'll have to standardize on `retval, err`, with `retval` being a table for forward compatibility. Of course, that's very restrictive...)

Reply:
>Also, I'm a bit confused, since we're commenting here on :EditQuery, which does not have a "caller" (in the sense of a calling function -- unless we are talking about the nvim_create_usercommand, which then needs to assert the call).

No, we're talking about `vim.treesitter.dev.edit_query()`, which does have a caller.

>If we want to promote this pattern, we'll have to standardize on retval, err, with retval being a table for forward compatibility. 

This is a good point, and I hadn't appreciated until this comment that the API in question (`edit_query`) does not return a value.

IMO for functions which are only called for their side effects (like this one), calling `assert` or `error` directly is appropriate (I realize that takes this entire discussion in a different direction, I apologize for not looking closer at this earlier).

The pattern of returning `retval, err` is the right one to use for functions which return a value (like `TSTreeView::new`).

Reply:
Thank you, that helps in wrapping my mind around this. It's clear that "one size fits all" fits none, so my goal is to understand (with the help of these examples) which pattern is appropriate in which situation; and the difference between "generators" (that return a value) and "actors" (that don't, and are called purely for their side effects) is one I tried making earlier.

What muddies the waters here unfortunately is that some of the functions change their type in these PRs...

Reply:
>  but I'm genuinely interested in having a discussion about this so I can understand the full pattern across our API (at least treesitter, possibly beyond). Ideally, we'll get something we can actually write in develop.txt and I can follow in future PRs instead of winging it based on contributor style as we (certainly I) have so far.

> (Part of the concern is that adding the error message to the return value locks the return part of the API. If we want to promote this pattern, we'll have to standardize on retval, err, with retval being a table for forward compatibility. Of course, that's very restrictive...)


I think part of the challenge here is that:

- there's no clear established pattern (yet)
- it's not one size fit all
- different contributors have slightly different preferences

From what I can tell we have at least:

1. `error("no luck")`
2. `retval, errmsg`
3. optional result, no errormsg
4. `on_error` parameter (async, visitors - where `n` errors can happen)
5. `vim.notify` (sometimes with optional `opts.silent` (async, visitors ^))


My personal reasoning for when to use which (so far):


1) is well suited for cases where the error is very exceptional, in a lower level and where the consumers are unlikely to handle it in a meaningful way. Advantage over 2) is that propagation happens for free and it's harder to accidentally swallow errors. (E.g. using `uv_handle/pipe:write()` without checking return values is common)

2) is well suited for cases where the error is more likely or where having a fallback is more realistic.

3) Kinda same as 2. where there is only a single case of "doesn't exist" (e.g. cache lookup, dict lookup)

4) lower level, either async methods or where the error can happen multiple times

5) Higher level. End-user invokes these directly

Reply:
Thank you, that is exactly the sort of guidance I was hoping for (and puts a proper face on my vague and confusing thoughts).

Reply:
> what would be guidelines for using pcall / assert? (Eating vs propagating vs. throwing errors).

I'd say:

`pcall` only if you either:

- rethrow and enrich the error with additional context
- can provide a fallback, or in case of a higher level function want to avoid the user getting a error with stacktrace but only a warning

`assert` is either:

- defensive programming: To make an assumption explicit that should always be true due to contracts.
- alternative to `error` ⬇️

In regards to `assert(truthy, msg)` vs. `if condition then error(msg) end`:
Use the latter if building the `msg` would be costly (and do it only within the if block. Otherwise `assert` is fine.


(Disclaimer: This is just how I've been using these things, not sure if there's consensus on these)

Reply:
Ah, that's a good point: `assert` needs a constructed message (which I believe should be the default ;))

Reply:
Incorporated @mfussenegger 's outline into https://github.com/neovim/neovim/pull/30240